ocaml/parsing/builtin_attributes.ml

```diff
@@ -808,43 +810,116 @@ let parse_ids_payload txt loc ~default ~empty cases payload =
       | Some r -> r
       | None -> warn ()
 
-let parse_property_attribute attr property =
+(* Looks for `arity n` in payload. If present, this returns `n` and an updated
+   payload with `arity n` removed *)
+let parse_arity payload =
```

Comment:
A lot of the grossness in this function seems to be the reconstruction of the new `payload` value to pass on to `parse_ids_payload`.

Suggestion: `parse_ids_payload` should take a function `string list -> 'a` as argument instead of an assoc list `(string list * 'a) list`. Then, the callsite of `parse_ids_payload` in `parse_property_attribute` has the flexibility to inspect the string list and filter out the arity-related pieces before attempting to match against the keys in the current assoc list.

The idea with this suggestion is that we can drop the code in `parse_arity` that reconstructs the payload, and instead cut it down to the `find_arity_in_args` core.

I recognize that this would mean refactoring `get_ids_from_exp` to also look for constants, or something like that. But this doesn't seem so bad to me.

Reply:
> I recognize that this would mean refactoring get_ids_from_exp to also look for constants, or something like that. But this doesn't seem so bad to me.

I would love to clean this up.  Are you proposing that `get_ids_from_exp` should just ignore constants entirely?  I think then we wouldn't warn if the payload has random other constants in it, or integers that aren't preceded by "arity".  And it seems like making it look explicitly for the "arity followed by int constant" pattern adds a similar level of complexity as we're eliminating in the other place.

Reply:
No, I'm suggesting `get_ids_and_constants_from_exp` that returns both ids and constants (both as strings).

> And it seems like making it look explicitly for the "arity followed by int constant" pattern adds a similar level of complexity as we're eliminating in the other place.

I think it's appreciably less as the suggestion allows us to avoid forging a `Pexp_apply` node / payload.

Reply:
(Btw I’m willing to be told that this isn’t worth it — feel free to resolve this and merge.)

Reply:
This didn't turn out to be quite so easy - there are problems with when we know the location and arity, needed to compute the lookup table (or parsing function) and warning message.  But, anyway, I did a slightly larger refactor somewhat along these lines that leads to much clearer code (and the need to forge this payload).